[WPE] WPE Platform: add clipboard support

carlosgcampos opened the pull request:
#### 6b152fb5e3abc63f49617726b453e32c39898d5b
<pre>
[WPE] WPE Platform: add clipboard support
<a href="https://bugs.webkit.org/show_bug.cgi?id=291075">https://bugs.webkit.org/show_bug.cgi?id=291075</a>

Reviewed by Patrick Griffis.

Add initial clipboard API with a default local only implementation and
support for text, markup and custom data.

* Source/WebKit/Shared/wpe/GRefPtrWPE.cpp:
(WTF::refGPtr):
(WTF::derefGPtr):
* Source/WebKit/Shared/wpe/GRefPtrWPE.h:
* Source/WebKit/UIProcess/API/wpe/PageClientImpl.cpp:
(WebKit::PageClientImpl::requestDOMPasteAccess):
* Source/WebKit/UIProcess/wpe/WebPasteboardProxyWPE.cpp:
(WebKit::clipboardFormats):
(WebKit::WebPasteboardProxy::getTypes):
(WebKit::WebPasteboardProxy::readText):
(WebKit::WebPasteboardProxy::readFilePaths):
(WebKit::WebPasteboardProxy::readBuffer):
(WebKit::WebPasteboardProxy::writeToClipboard):
(WebKit::WebPasteboardProxy::clearClipboard):
(WebKit::WebPasteboardProxy::typesSafeForDOMToReadAndWrite):
(WebKit::WebPasteboardProxy::writeCustomData):
(WebKit::pasteboardItemInfoFromFormats):
(WebKit::WebPasteboardProxy::allPasteboardItemInfo):
(WebKit::WebPasteboardProxy::informationForItemAtIndex):
(WebKit::WebPasteboardProxy::getPasteboardItemsCount):
(WebKit::WebPasteboardProxy::readURLFromPasteboard):
(WebKit::WebPasteboardProxy::readBufferFromPasteboard):
(WebKit::WebPasteboardProxy::getPasteboardChangeCount):
* Source/WebKit/WPEPlatform/CMakeLists.txt:
* Source/WebKit/WPEPlatform/wpe/WPEClipboard.cpp: Added.
(wpeClipboardDispose):
(wpeClipboardSetProperty):
(wpeClipboardGetProperty):
(wpeClipboardChanged):
(wpe_clipboard_class_init):
(wpe_clipboard_new):
(wpe_clipboard_get_display):
(wpe_clipboard_get_change_count):
(wpe_clipboard_get_formats):
(wpe_clipboard_set_content):
(wpe_clipboard_get_content):
(wpe_clipboard_read_bytes):
(wpe_clipboard_read_text):
(wpe_clipboard_content_new):
(wpe_clipboard_content_ref):
(wpe_clipboard_content_unref):
(wpe_clipboard_content_set_text):
(wpe_clipboard_content_set_markup):
(wpe_clipboard_content_set_bytes):
(wpe_clipboard_content_serialize):
* Source/WebKit/WPEPlatform/wpe/WPEClipboard.h: Added.
* Source/WebKit/WPEPlatform/wpe/WPEDisplay.cpp:
(wpe_display_get_clipboard):
* Source/WebKit/WPEPlatform/wpe/WPEDisplay.h:
* Source/WebKit/WPEPlatform/wpe/wpe-platform.h:
* Tools/WebKitTestRunner/wpe/UIScriptControllerWPE.cpp:
(WTR::UIScriptControllerWPE::copyText):

Canonical link: <a href="https://commits.webkit.org/294505@main">https://commits.webkit.org/294505@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fd9676ec969174cc5df7b47e6aed6e76bcf6e1d9

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/101494 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/21160 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/11488 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/106648 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/52123 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/103534 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/21467 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/29654 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/77298 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/34325 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/104501 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/16566 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/91670 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/57640 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/16388 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/9693 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/51471 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/86256 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/9774 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/108999 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/28623 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/21056 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/86264 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/28985 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/87877 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/85831 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/30566 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/8291 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/22749 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/16593 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/28554 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/33835 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/28365 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/31685 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/29924 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

TingPing commented on Source/WebKit/WPEPlatform/wpe/WPEClipboard.cpp:
```diff
@@ -0,0 +1,464 @@
+    std::optional<CString> text;
+    std::optional<CString> markup;
+    std::optional<HashMap<const char*, GRefPtr<GBytes>>> buffers;
+    int referenceCount { 1 };
```
Glib has API for this https://docs.gtk.org/glib/reference-counting.html

carlosgcampos replied:
Yes, I know, but we still want to use fast malloc

TingPing reviewed (approved):
Mentioned elsewhere, but I'd like all of our clipboard APIs to be async to match the reality of the platform. But I understand those are deeper WebKit changes.

carlosgcampos committed:
[WPE] WPE Platform: add clipboard support
https://bugs.webkit.org/show_bug.cgi?id=291075

Reviewed by Patrick Griffis.

Add initial clipboard API with a default local only implementation and
support for text, markup and custom data.

* Source/WebKit/Shared/wpe/GRefPtrWPE.cpp:
(WTF::refGPtr):
(WTF::derefGPtr):
* Source/WebKit/Shared/wpe/GRefPtrWPE.h:
* Source/WebKit/UIProcess/API/wpe/PageClientImpl.cpp:
(WebKit::PageClientImpl::requestDOMPasteAccess):
* Source/WebKit/UIProcess/wpe/WebPasteboardProxyWPE.cpp:
(WebKit::clipboardFormats):
(WebKit::WebPasteboardProxy::getTypes):
(WebKit::WebPasteboardProxy::readText):
(WebKit::WebPasteboardProxy::readFilePaths):
(WebKit::WebPasteboardProxy::readBuffer):
(WebKit::WebPasteboardProxy::writeToClipboard):
(WebKit::WebPasteboardProxy::clearClipboard):
(WebKit::WebPasteboardProxy::typesSafeForDOMToReadAndWrite):
(WebKit::WebPasteboardProxy::writeCustomData):
(WebKit::pasteboardItemInfoFromFormats):
(WebKit::WebPasteboardProxy::allPasteboardItemInfo):
(WebKit::WebPasteboardProxy::informationForItemAtIndex):
(WebKit::WebPasteboardProxy::getPasteboardItemsCount):
(WebKit::WebPasteboardProxy::readURLFromPasteboard):
(WebKit::WebPasteboardProxy::readBufferFromPasteboard):
(WebKit::WebPasteboardProxy::getPasteboardChangeCount):
* Source/WebKit/WPEPlatform/CMakeLists.txt:
* Source/WebKit/WPEPlatform/wpe/WPEClipboard.cpp: Added.
(wpeClipboardDispose):
(wpeClipboardSetProperty):
(wpeClipboardGetProperty):
(wpeClipboardChanged):
(wpe_clipboard_class_init):
(wpe_clipboard_new):
(wpe_clipboard_get_display):
(wpe_clipboard_get_change_count):
(wpe_clipboard_get_formats):
(wpe_clipboard_set_content):
(wpe_clipboard_get_content):
(wpe_clipboard_read_bytes):
(wpe_clipboard_read_text):
(wpe_clipboard_content_new):
(wpe_clipboard_content_ref):
(wpe_clipboard_content_unref):
(wpe_clipboard_content_set_text):
(wpe_clipboard_content_set_markup):
(wpe_clipboard_content_set_bytes):
(wpe_clipboard_content_serialize):
* Source/WebKit/WPEPlatform/wpe/WPEClipboard.h: Added.
* Source/WebKit/WPEPlatform/wpe/WPEDisplay.cpp:
(wpe_display_get_clipboard):
* Source/WebKit/WPEPlatform/wpe/WPEDisplay.h:
* Source/WebKit/WPEPlatform/wpe/wpe-platform.h:
* Tools/WebKitTestRunner/wpe/UIScriptControllerWPE.cpp:
(WTR::UIScriptControllerWPE::copyText):

Canonical link: https://commits.webkit.org/294505@main
